### ISSUE TITLE
aws_batch_compute_environment: problems updating/changing tags

### ISSUE THREAD

**Issue by davekaufman** [terraform-providers/terraform-provider-aws#2696]
### Terraform Version
```terraform -v                                                                                                                                                       9:38
Terraform v0.11.1
+ provider.archive v1.0.0
+ provider.aws v1.5.0
+ provider.template v1.0.0
```

### Affected Resource(s)
- aws_batch_compute_environment

### Terraform Configuration Files
```resource "aws_batch_compute_environment" "test-batch-compute" {
  compute_environment_name = "test-batch-compute"
  compute_resources {
    bid_percentage      = 50
    ec2_key_pair        = "${var.keyname}"
    image_id            = "${var.ami}"
    instance_role       = "${data.aws_iam_role.innstance_role.arn}"
    instance_type       = [ "optimal" ]
    max_vcpus           = 256
    min_vcpus           = 0
    security_group_ids  = [ "${data.aws_security_group.batch_sg.id}" ]
    spot_iam_fleet_role = "${data.aws_iam_role.aws-ec2-spot-fleet-role.arn}"
    subnets             = [ "${data.aws_subnet.us-east-1d.id}", "${data.aws_subnet.us-east-1e.id}" ]
    tags {
      "Name"            = "test-batch-compute"
      "NewTag"        = "New"
    }
    type                = "SPOT"
  }
  service_role          = "${data.aws_iam_role.AWSBatchServiceRole.arn}"
  type                  = "MANAGED"
}

```

### Expected Behavior
Updating the tags in an aws_batch_compute_environment definition results in error and then discrepancy between actual tags and terraform state.

To reproduce, using above: add `NewTag` definition to above, run `terraform apply`.  Expect number of tags to now be 2.

### Actual Behavior
Get Error on first `terraform apply`:
```aws_batch_compute_environment.test-batch-compute: Modifying... (ID: test-batch-compute)
  compute_resources.0.tags.%:       "1" => "2"
  compute_resources.0.tags.NewTag: "" => "New"

Error: Error applying plan:

1 error(s) occurred:

* aws_batch_compute_environment.test-batch-compute: 1 error(s) occurred:

* aws_batch_compute_environment.test-batch-compute: : Manually scaling down compute environment is not supported.  Disconnecting job queues from compute environment will cause it to scale-down to minvCpus.
        status code: 400, request id: bd28eaf0-e419-11e7-96f7-15ecbe57f6bf
```
Re-run `terraform apply`, changing nothing.  
`Apply complete! Resources: 0 added, 0 changed, 0 destroyed.`  Number of tags shown by `aws batch describe-compute-environments` does not change (still only 1, in this example), and does not match the number terraform thinks is present (2).

### Steps to Reproduce
Change number of tags in compute environment definition (either add new or remove existing). Run `terraform apply`.  Get error about scaling compute env. Re-run after scaling error and get the standard nothing to change message.

### Important Factoids
None

### References
None

**Comment by rajbettaswamy** [comment]
Even I am facing the same issue now. I think this is because of how we have pre-defined ` min_vcpus` 
 to `0` for that respective `aws_batch_compute_environment`.

At the time TF run, I have couple of job in submitted pipeline and during that time AWS Batch really can't scale down because of already running EC2 vCPUs. So, the `terraform apply` should work in next subsequent run when there is no submitted job in the pipeline.

Is it possible to make defining `min_vcpus` as *optional* and no default values can be assigned to it? Or  Can we fetch the value for `min_vcpus` dynamically depending upon the number of currently running EC2 vCPUs using that compute environment ?

**Comment by bflad** [comment]
Hi folks 👋 Thank you for reporting this issue. This issue does appear to be valid, in the sense that Terraform is claiming it can update compute resource tags in place, however the Batch API does not support this: https://docs.aws.amazon.com/batch/latest/APIReference/API_UpdateComputeEnvironment.html

We will fix the resource schema to have `ForceNew: true` for the `tags` argument within the `compute_resources` configuration block so Terraform will propose recreating the `aws_batch_compute_environment` resource if the tags are modified instead of proposing an update which it cannot perform. This issue will be closed with that fix.

As for the enhancement to support updating compute resource tags, it will need to be implemented in the Batch API before it can be implemented in Terraform. We suggest creating an AWS Support case as a feature request or reaching out to your Technical Account Manager if you have one. 👍

**Comment by bflad** [comment]
The Terraform update [mentioned above](https://github.com/terraform-providers/terraform-provider-aws/issues/2696#issuecomment-555690931) has been merged and will release with version 2.39.0 of the Terraform AWS Provider, later today. Please also see the end of that comment for information about supporting in-place updates in the future.
